consensus/src/consensus_observer/pending_blocks.rs

```diff
@@ -5,98 +5,205 @@ use crate::consensus_observer::{
     logging::{LogEntry, LogSchema},
     network_message::OrderedBlock,
 };
+use aptos_config::config::ConsensusObserverConfig;
 use aptos_consensus_types::{common::Round, pipeline::commit_decision::CommitDecision};
 use aptos_infallible::Mutex;
-use aptos_logger::debug;
-use aptos_types::{block_info::BlockInfo, ledger_info::LedgerInfoWithSignatures};
+use aptos_logger::{debug, error, warn};
+use aptos_types::{
+    block_info::BlockInfo, epoch_change::Verifier, epoch_state::EpochState,
+    ledger_info::LedgerInfoWithSignatures,
+};
 use std::{collections::BTreeMap, sync::Arc};
 
 /// A simple struct to store the block payloads of ordered and committed blocks
 #[derive(Clone)]
 pub struct PendingOrderedBlocks {
-    // Pending ordered blocks (indexed by consensus round)
-    pending_ordered_blocks: Arc<Mutex<BTreeMap<Round, (OrderedBlock, Option<CommitDecision>)>>>,
+    // The configuration of the consensus observer
+    consensus_observer_config: ConsensusObserverConfig,
+
+    // Verified pending ordered blocks (for the current epoch)
+    verified_pending_blocks:
+        Arc<Mutex<BTreeMap<(u64, Round), (OrderedBlock, Option<CommitDecision>)>>>,
+
+    // Unverified pending ordered blocks (for future epochs)
+    unverified_pending_blocks:
+        Arc<Mutex<BTreeMap<(u64, Round), (OrderedBlock, Option<CommitDecision>)>>>,
```

Comment:
Instead of two Maps and moving data between then, can we have an Enum of Verified(OrderedBlock) and Unverified(OrderedBlock) as values? That way it's less prone to deadlocks now and in future.

Reply:
SGTM (I've already done this locally, but I'll just merge it into this PR for now! 😄)